perf: optimise composer send behaviour

hurali97 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This fixes major things wrong with `ReportScreen` and it's children, `ReportActionCompose` and `ComposerWithSuggestions`. The highlight is the `personalDetails` prop from `onyx` as on high traffic accounts `personalDetails` becomes a huge object and is expensive to pass as a prop and have the component render/ re-render. So each time a message is sent, we are manipulating onyx for the  `report_`, `reportActions_` and `personalDetails` keys first with the optimistic data and then with the successful data. Most of the `ReportScreen` and it's children doesn't really need `personalDetails` as a prop from onyx as they are passing that prop to certain utility functions to calculate for example, icon. So this `personalDetails` prop can be removed and we can use directly from the utility as we have `Onyx.connect` already present in the utility files.

Apart from this, some props including `report` and `reportActions` have been simplified to become lightweight so that the can be passed easily down to the children. Previously, we were passing the whole object and it had properties that we don't need. Also, to prevent unnecessary re-rendering of `ReportActionCompose` the empty state is lifted and moved to `ReportActionCompose` alongside the `SendButton` and `EmojiButton`. Also, `sortedReportActions` aren't really needed to pass to `ReportActionsListItemRenderer` as it uses the prop to only evaluate whether it's a group or not. So we have passed a new prop `displayAsGroup` to the component directly instead of passing the whole actions array.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/26347
PROPOSAL: https://github.com/Expensify/App/issues/26347#issuecomment-1748791267


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

#### Testing Steps:

- Open a Report and focus the composer
- Start typing some letters and then hit send
- Once the message is sent, start typing again and then press send
- No message should be present in the composer and the behaviour should feel smooth

- [X] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47336142/f12e7031-3fd1-4576-ae48-9c3ba1d94f39

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47336142/7a108c17-d5ef-4eb3-a904-b123875693dc

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47336142/ec878cfc-f2ff-49f6-9f2d-19ff7d3b9b7d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47336142/43ab0dc9-373d-4226-ac92-fdd7cea21384

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47336142/b6f81d25-d62f-4de5-bb76-9205ee99731a

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/47336142/983298cb-5e53-4921-b661-4bf58375f451

</details>

hurali97 committed:
perf: memoize components

hurali97 committed:
refactor: simplify props

hurali97 committed:
perf: reduce re-renders

hurali97 committed:
refactor: make action prop lightweight

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
perf: add Interaction Manager

hurali97 committed:
perf: reference bindings of functions

hurali97 committed:
perf: optimise rendering

hurali97 committed:
perf: optimise usage of personal detail list

hurali97 committed:
fix: linting

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
refactor: remove not needed code

hurali97 committed:
fix: linting and test

hurali97 committed:
fix: failing unread indicator test

hurali97 committed:
fix: prettier issues

hurali97 committed:
fix: failing test

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: reassure failures reported

hurali97 commented:
Figuring out why Reassure tests are failing and will update the PR accordingly.

hurali97 committed:
fix: reported reassure issues

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: reported reassure issues

hurali97 commented:
@aimane-chnaif All green now !

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

aimane-chnaif reviewed (commented):
After this refactor, https://github.com/Expensify/App/issues/26347 issue still remaining right?

hurali97 commented:
> After this refactor, #26347 issue still remaining right?

@aimane-chnaif No this will fix that issue because the refactor happened following that issue.

I will take a look at your comments shortly.

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented on src/pages/home/ReportScreen.js:
```diff
@@ -499,14 +480,31 @@ export default compose(
     ),
-)(ReportScreen);
+)(
+    memo(
+        ReportScreen,
+        (prevProps, nextProps) =>
+            prevProps.isSidebarLoaded === nextProps.isSidebarLoaded &&
+            _.isEqual(prevProps.reportActions, nextProps.reportActions) &&
+            _.isEqual(prevProps.reportMetadata, nextProps.reportMetadata) &&
+            prevProps.isComposerFullSize === nextProps.isComposerFullSize &&
+            _.isEqual(prevProps.betas, nextProps.betas) &&
+            _.isEqual(prevProps.policies, nextProps.policies) &&
+            prevProps.accountManagerReportID === nextProps.accountManagerReportID &&
+            _.isEqual(prevProps.personalDetails, nextProps.personalDetails) &&
+            prevProps.userLeavingStatus === nextProps.userLeavingStatus &&
+            prevProps.report.reportID === nextProps.report.reportID &&
+            prevProps.report.policyID === nextProps.report.policyID &&
+            prevProps.report.isOptimisticReport === nextProps.report.isOptimisticReport &&
+            prevProps.report.statusNum === nextProps.report.statusNum &&
+            _.isEqual(prevProps.report.pendingFields, nextProps.report.pendingFields) &&
+            prevProps.currentReportID === nextProps.currentReportID,
+    ),
```
I agree and I tested it on all of the devices using same and different accounts !!

hurali97 committed:
fix: pr review

hurali97 commented:
@aimane-chnaif Are you able to review the PR again? 🙂

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif Resolved conflicts !

hurali97 commented:
> > > After this refactor, #26347 issue still remaining right?
> > 
> > 
> > @aimane-chnaif No this will fix that issue because the refactor happened following that issue.
> 
> I am asking because after code review, this PR still doesn't fix the root cause of #26347. I tested on android emulator:
> 
>  Screen.Recording.2023-11-15.at.2.28.35.PM.mov


@aimane-chnaif Can we try to test this on a real world scenario? For example, test it on a device since there's no connected keyboard there and the reason we see it in emulator is because we can type really quickly directly from the connected keyboard which is not a real world scenario.

I tested it on android device and it works greatly. Let me know if you face the same on real android device.

hurali97 commented on src/components/AvatarWithDisplayName.js:
```diff
@@ -93,8 +88,8 @@ function AvatarWithDisplayName(props) {
     const subtitle = ReportUtils.getChatRoomSubtitle(props.report);
     const parentNavigationSubtitleData = ReportUtils.getParentNavigationSubtitle(props.report);
     const isMoneyRequestOrReport = ReportUtils.isMoneyRequestReport(props.report) || ReportUtils.isMoneyRequest(props.report);
-    const icons = ReportUtils.getIcons(props.report, props.personalDetails, props.policy);
```
Yeah that's why I was surprised 😄 So I have supplied the appropriate params now, I think !

hurali97 committed:
fix: add default value

mountiny commented:
running it although there is a bunch of conflicts https://github.com/Expensify/App/actions/runs/6894604269

OSBotify commented:
:test_tube::test_tube: Use the links below to test this adhoc build on Android, iOS, Desktop, and Web. Happy testing! :test_tube::test_tube:
| Android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/30168/index.html  | https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/30168/index.html  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/30168/index.html)  | ![iOS](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/ios/30168/index.html)  |
| Desktop :computer: | Web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/30168/NewExpensify.dmg  | https://30168.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/30168/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://30168.pr-testing.expensify.com)  |

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif Conflicts resolved.

However, on one of my mid-lower end device I could repro this again on Prod build downloaded by QR. On my high end device, it wasn't reproducible.

https://github.com/Expensify/App/assets/47336142/af8e5943-7f6f-4291-a6d8-1039c64df272

Although this PR still fixes a lot of other performance issues especially immediately toggling the Send button when user starts to type. I would still want to investigate further on my mid-lower end device to see why the compose is not getting empty.

Here's the catch, I am going on vacation for the next week and will be available after that, so I can continue work on this. Till then, we can merge this PR and have someone else continue the work or I can continue the work when I am back. Let me know what fits the best !

mountiny commented:
I don't think this pr should have solved the composer sent issue so that is fine as long as it does not make it worse

@aimane-chnaif has other testing been good? Can we go ahead with a merge?

mountiny commented:
@hurali97 could you please sync up with main to resolve conflicts? thank you!

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif conflicts resolved 👍

OSBotify commented:
:test_tube::test_tube: Use the links below to test this adhoc build on Android, iOS, Desktop, and Web. Happy testing! :test_tube::test_tube:
| Android :robot:  | iOS :apple: |
| ------------- | ------------- |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/android/30168/index.html  | ❌ FAILED ❌  |
| ![Android](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/android/30168/index.html)  | The QR code can't be generated, because the iOS build failed  |
| Desktop :computer: | Web :spider_web: |
| https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/30168/NewExpensify.dmg  | https://30168.pr-testing.expensify.com  |
| ![Desktop](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://ad-hoc-expensify-cash.s3.amazonaws.com/desktop/30168/NewExpensify.dmg)  | ![Web](https://api.qrserver.com/v1/create-qr-code/?size=120x120&data=https://30168.pr-testing.expensify.com)  |

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif Both of the screenshots look the same to me. Can you elaborate more ? 🙂

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: appropriate arguments to getDisplayNameOrDefault

hurali97 committed:
fix: workspace unavailable avatar

hurali97 commented:
@aimane-chnaif It's fixed now. We are passing the wrong order of parameters to `getIcons` in main but it worked and wasn't caught. Anyways, I fixed it now with retaining correct order of parameters.

<img width="1252" alt="Screenshot 2023-11-30 at 12 21 24 PM" src="https://github.com/Expensify/App/assets/47336142/6ccae323-eb6f-4fed-a5bb-f4deae58327e">

hurali97 commented on src/pages/ReportDetailsPage.js:
```diff
@@ -254,7 +254,7 @@ export default compose(
             key: ONYXKEYS.PERSONAL_DETAILS_LIST,
         },
         policies: {
```
Oh yeah, correct !!

hurali97 committed:
fix: pass policy instead of policies to getIcons

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
test: fix failing caledarpickertest due to an edge case of last month in the calendar

hurali97 commented on tests/unit/CalendarPickerTest.js:
```diff
@@ -64,7 +64,20 @@ describe('CalendarPicker', () => {
         fireEvent.press(getByTestId('next-month-arrow'));
 
-        const nextMonth = new Date().getMonth() + 1;
+        /**
+         * Handle edge case for last month "Decemeber" as
+         * adding a 1 to it will produce an invalid month
+         * 12 since the last index supported is 11, so we
+         * set it to 0 to represent January.
+         */
+        let nextMonth = new Date().getMonth();
+
+        if (nextMonth !== 11) {
+            nextMonth += 1;
+        } else {
+            nextMonth = 0;
+        }
+
```
I see, will do that once it's landed on main 👍

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif Okay I will revert in those places where profile picture is not being updated 👍 

Can you tell me also, when I update profile photo on mobile, is it supposed to be changed in real-time on web also? Like as you have opened the two in the video?

hurali97 committed:
fix: personal details not updating

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif So on main there's also no real-time update of profile details. Refer to my below video where I have to change the page in order for refresh to happen. 

However on my PR as you reported, I fixed the issue related to profile picture not updating on android. Also in the header on web, the profile picture wasn't being updated even after changing the page, so I have fixed that as well.

### main:

https://github.com/Expensify/App/assets/47336142/1dbbc338-7cdc-417b-ac6d-7b0c7852bda7

### This PR:


https://github.com/Expensify/App/assets/47336142/771f2ee0-9d23-4553-8784-b0463dd9a099

hurali97 commented:
@aimane-chnaif Thanks for explaining 👍 I will test this again !

hurali97 commented:
@aimane-chnaif Okay so it works for me on main but not in real-time per se but after 25-30 sec update is reflected on the second account but I guess that might be due to a certain reasons. 

However, as you can see in video this specific little icon doesn't update after 30 seconds or even if I switch chat, so this is probably a bug specific to this little icon, in main.


https://github.com/Expensify/App/assets/47336142/ff57b127-0b47-4a22-bfc0-f48db1f99fd3

I will test this now on my PR and fix the correspondence.

hurali97 committed:
fix: personalDetails not updating

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: ts issues

hurali97 commented:
@aimane-chnaif So I have fixed the `personalDetails` not being updated on my PR and also fixed the issue that's present on main which doesn't update the little icon in LHN.

On both main and this PR, update takes around +-30ms to reflect, probably because I am on debug or my internet connection.

https://github.com/Expensify/App/assets/47336142/70bf4e4a-3583-4987-9ba6-1ba10bf012d7

hurali97 commented:
@aimane-chnaif Sure thing, I will do that today !!

hurali97 committed:
fix: personaldetails not updating

hurali97 committed:
fix: personalDetails not updating in AnonymousReportFooter

hurali97 committed:
refactor: cleanup

hurali97 committed:
refactor: cleanup

hurali97 committed:
refactor: move personalDetails to LHNOptionsList

hurali97 committed:
fix: typescript issues

hurali97 commented:
@aimane-chnaif are you able to test again? I tried to test myself as much as I know about the app and I think it looks good now 👍

aimane-chnaif reviewed (commented):
I keep testing all affected pages

hurali97 commented on src/components/AnonymousReportFooter.tsx:
```diff
@@ -30,9 +33,9 @@ function AnonymousReportFooter({isSmallSizeLayout = false, personalDetails, repo
-                    personalDetails={personalDetails}
                     isAnonymous
                     shouldEnableDetailPageNavigation
+                    policy={policy}
```
I can explain. Previously, we were passing the `policies` prop to `ReportFooter` and it has `ReportAnonymousFooter` as a child so it re-renders whenever `policies` update. But since I removed that altogether, it was not updating in real-time anymore. So I decided to use the dedicated `policy` in the nearest component possible.

hurali97 commented on src/components/AvatarWithDisplayName.tsx:
```diff
@@ -22,6 +22,9 @@ import Text from './Text';
     parentReportActions: OnyxEntry<ReportActions>;
+
+    /** Personal details of all users */
+    personalDetails: OnyxEntry<Record<string, PersonalDetails>>;
```
I am not sure of this, I followed this approach from the other parts of the codebase.

hurali97 commented on src/components/AttachmentModal.js:
```diff
@@ -394,6 +394,14 @@ function AttachmentModal(props) {
+        }),
+        [validateAndDisplayFileToUpload],
+    );
+
```
This change is basically to memoize `AttachmentModal` and avoid re-rendering `ComposerWithSuggestions` when not required. Hence I have also moved `ComposerWithSuggestions` to not remain a children of `AttachmentModal`.

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

mountiny commented:
@aimane-chnaif lets focus on this one please 🥺

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented on src/pages/home/ReportScreen.js:
```diff
@@ -166,16 +161,24 @@ function ReportScreen({
+    const isEmptyChat = useMemo(() => _.isEmpty(reportActions), [reportActions]);
     // There are no reportActions at all to display and we are still in the process of loading the next set of actions.
-    const isLoadingInitialReportActions = _.isEmpty(reportActions) && reportMetadata.isLoadingInitialReportActions;
+    const isLoadingInitialReportActions = isEmptyChat && reportMetadata.isLoadingReportActions;
```
This `isLoadingInitialReportActions` is already present in main and I only changed the redundant `_.isEmpty` condition and moved it to a separate variable with memo `isEmptyChat`. I haven't introduced this `isLoadingInitialReportActions ` boolean and is something that main already contans.

hurali97 replied:
Oh that... it's resolved now, I think it might be a typo from me but it was fixed after pulling in the latest main and resolving the conflicts. Thanks for flagging !

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif done, thanks 🙂

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: typecheck for AvatarWithDisplayName

hurali97 committed:
fix: linting

hurali97 commented:
@aimane-chnaif any leads on the testing? 🙂

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

mountiny commented:
@aimane-chnaif Can you resolve conflicts locally and finish the review here please? thank you!

mountiny commented:
@allroundexperts Has approved the PR here https://github.com/Expensify/App/pull/32336 which is build atop of this one so I would really love to get this one going

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

mountiny commented:
@hurali97 Another conflicts

@aimane-chnaif could you approve the PR once all is tested?

hurali97 commented:
Coming from [here](https://callstack-hq.slack.com/archives/C05LX9D6E07/p1703844985106289?thread_ts=1703679928.194659&cid=C05LX9D6E07). Waiting on to fix Reassure Tests.

In the mean time, I will keep resolving conflicts so that we don't have huge conflicts at the end, when Reassure tests are fixed and we are following up again on this PR. Thanks.

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

mountiny commented:
Restarting the reassure tests since this PR got merged https://github.com/Expensify/App/pull/33831

ArekChr commented:
@hurali97 please update the branch with the latest main to sync new perf-test gh action

mountiny commented:
Yep, my bad, though this would be taking the action from main

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif Conflicts resolved !

mountiny commented:
> sorry there's another conflict, while I almost completed testing

I dont think this matters so much, lets assume the conflict is simple, can you post the results of your testing now please?

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
> BUG: [web] Copy + paste image to composer doesn't work
> 
> <img alt="Screenshot 2024-01-10 at 12 20 54 PM" width="620" src="https://private-user-images.githubusercontent.com/96077027/295548327-470d751f-9600-4181-bb14-140a4cd26509.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4ODYxNTcsIm5iZiI6MTcwNDg4NTg1NywicGF0aCI6Ii85NjA3NzAyNy8yOTU1NDgzMjctNDcwZDc1MWYtOTYwMC00MTgxLWJiMTQtMTQwYTRjZDI2NTA5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDExMjQxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiMjY3NmUzNDllZWRiYzZjOGVlN2RhMTViNDkwZDJjMmEzOGRhNzNmMWQzNjVlMzg0MDM1ZTZiZTliNzhjNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.cmNzUl_m7F0qQvfrksIwx7DszeteAMAjhRCszVyUcf8">


How can I reproduce this? By copying an image and pasting it in compose input?

mountiny commented:
Thanks @aimane-chnaif

hurali97 committed:
fix: paste not working on web

hurali97 commented:
@aimane-chnaif It's fixed now 👍 

https://github.com/Expensify/App/assets/47336142/fe8eaedf-19af-47d0-bf73-e6fd18398bf4

hurali97 commented:
@aimane-chnaif  can you verify again? I have tried to reproduce this on main and it's reproducible on main. I recall there was some change merged to main a week ago, where this functionality was removed from `onSubmitComment`. 

See [here](https://github.com/Expensify/App/blob/1f7a97b59e269f974bf3823bb079d89f28f7f995/src/pages/home/ReportScreen.js#L329C1-L338C7) there's no `scrollToOffset` anymore.

https://github.com/Expensify/App/assets/47336142/cc399a4f-9406-40bd-9310-6858a0c06e5e

hurali97 committed:
fix: compose height on mobile

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

mountiny commented on src/styles/index.ts:
```diff
@@ -1885,6 +1885,14 @@ const styles = (theme: ThemeColors) =>
             height: 40,
             padding: 10,
             margin: 3,
+            ...Platform.select({
```
This is against our guidelines 😬

hurali97 replied:
> This is against our guidelines 😬

Oh I apologize. 

> what's the root cause?

Basically, I moved the `EmojiButton` and `SendButton` from `ReportActionCompose` to `ComposerWithSuggestions` and then it started to happen. The reason was to keep the re-rendering in one place only and don't re-render `ReportActionCompose`. It was basically concerning the empty state, which was previously defined in `ReportActionCompose`.

 Can you guys point me where can I look for design guidelines? OR anything relevant.

hurali97 replied:
Thanks, it seems quite complicated to move `SendButton` and `EmojiButton` to platform specific files and then apply the style there. I will see if I can find a way where we have to make less changes to the codebase,

hurali97 replied:
Alright, it's fixed now. I opted for a simpler approach and that's to take out `SendButton` and `EmojiButton` outside of the `View` encapsulating `Composer` in `ComposeWithSuggestions`.

hurali97 committed:
fix: revert platform specific styles

hurali97 committed:
fix: extra compose height

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

mountiny commented:
@aimane-chnaif is going through the testing now

hurali97 commented:
@aimane-chnaif Have you completed your testing ? Are there anymore bugs ?

hurali97 commented:
> No more bugs

Great, I will proceed with fixing the Reply bug tomorrow morning 👍

hurali97 committed:
fix: reply not shown

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
https://github.com/Expensify/App/assets/47336142/1f38b61c-0f12-409a-a5f2-1638813b94fd

@aimane-chnaif Reply bug is fixed 👍

hurali97 committed:
fix: add missing reportAction properties

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented on src/pages/home/report/ReportActionsListItemRenderer.js:
```diff
@@ -55,6 +55,49 @@ function ReportActionsListItemRenderer({
+            created: reportAction.created,
+            actorAccountID: reportAction.actorAccountID,
+            childVisibleActionCount: reportAction.childVisibleActionCount,
+            childOldestFourAccountIDs: reportAction.childOldestFourAccountIDs,
```
Thanks for flagging this. Done 👍

hurali97 commented on src/pages/home/report/ReportActionsListItemRenderer.js:
```diff
@@ -55,6 +55,71 @@ function ReportActionsListItemRenderer({
+            reportAction.childStatusNum,
+            reportAction.childReportName,
+            reportAction.childManagerAccountID,
+            reportAction.childMoneyRequestCount,
```
Here's the catch. Right now each update to Onyx collection result in a new object instance creation which breaks the memoization itself. So to fix it better, we can just pick what we need and construct a memoized object which will only update when a property changes.

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif I don't understand the issue clearly. Looks like main branch is not showing the Chat's welcome message and this branch is showing it. This looks like the ideal behaviour and maybe something recently has been merged to main?

hurali97 committed:
fix: hide reply and task welcome message and add the missing fields

hurali97 committed:
refactor: remove not used prop

hurali97 commented:
https://github.com/Expensify/App/assets/47336142/1e4a2585-756b-49f3-9279-fb0465d9fb74

@aimane-chnaif fixed the issue and also double-checked any missing properties.

hurali97 commented:
> I did full regression test as much as what QA team does 😄 Hopefully above bugs are the last ones 🙏

Thanks, I appreciate your effort 🙂 . I will let you know once I fix these 👍

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: local time not displayed

hurali97 committed:
refactor: move memoized report object to report screen

hurali97 commented:
@aimane-chnaif I have fixed the above issues except for the composer focus [one](https://github.com/Expensify/App/pull/30168#issuecomment-1899887260). Sometimes it focuses and sometimes not, so I am having hard time figuring it out. I will focus on this bug on Monday morning. 👍

hurali97 committed:
fix: composer not being auto focused

hurali97 committed:
refactor: remove not needed code change

hurali97 committed:
fix: linting

hurali97 committed:
fix: reassure tests

hurali97 committed:
refactor: add prop types

mountiny commented:
@hurali97 more conflicts

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif It's ready 👍

mountiny commented:
https://github.com/Expensify/App/actions/runs/7656273034

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif What's the proceedings now? 

TS checks and Linting is failing after pulling in the latest main, so I can't fix it.

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 committed:
fix: crash and profile not updating

hurali97 committed:
Merge branch 'main' of https://github.com/hurali97/Expensify-App into perf/optimise-composer-send

hurali97 commented:
@aimane-chnaif Thanks for doing a regression. I have fixed these issues now 👍

aimane-chnaif reviewed (approved):
I think we can now merge this as no bugs found at this moment from our side.
But still be on a look out for any regressions QA is reporting since this is big refactor.

mountiny reviewed (approved):
Let's go then, code looks good to me, this was a massive effort, thanks everyone!

mountiny reviewed (approved):
Let's go then, code looks good to me, this was a massive effort, thanks everyone!

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7733329028) to staging by https://github.com/mountiny in version: 1.4.35-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

rlinoz commented:
I believe this PR caused this regression: https://github.com/Expensify/App/issues/35559

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7757757234) to production by https://github.com/mountiny in version: 1.4.35-7 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

tgolen commented on src/pages/home/report/ReportActionsListItemRenderer.js:
```diff
@@ -55,6 +55,71 @@ function ReportActionsListItemRenderer({
+            childVisibleActionCount: reportAction.childVisibleActionCount,
+            childOldestFourAccountIDs: reportAction.childOldestFourAccountIDs,
+            childType: reportAction.childType,
+            person: reportAction.person,
```
Is this referenced somehere? I can't find it anywhere.

mountiny replied:
Ah that would be for threads, it would have to be used for that little element showing how many replies are in thread, who responded, right?
